HW2/P3/mandelbrot.pyx

```diff
@@ -9,39 +9,158 @@ from cython.parallel import prange
-                c = in_coords[i, j]
-                z = 0
-                for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
-
+    assert in_coords.shape[0] == out_counts.shape[
+        0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[
+        1],  "Input and output arrays must be the same size"
+
+    # get real / imag parts
+    real_coords = np.real(in_coords)
+    imag_coords = np.imag(in_coords)
+
+    # version multithreading and AVX instructions...
+    for i in prange(in_coords.shape[0], nogil=True, num_threads=4, schedule='static', chunksize=1):
+      for j in range(in_coords.shape[1] / 8):
+
+        # parallelize 8 complex numbers
+        # first separate real and imaginary parts
+        realc = real_coords[i, j*8]
+        imagc = imag_coords[i, j*8]
+
+        avxrealc = AVX.make_float8(real_coords[i, j * 8 + 7],
+                                   real_coords[i, j * 8 + 6],
+                                   real_coords[i, j * 8 + 5],
+                                   real_coords[i, j * 8 + 4],
+                                   real_coords[i, j * 8 + 3],
+                                   real_coords[i, j * 8 + 2],
+                                   real_coords[i, j * 8 + 1],
+                                   real_coords[i, j * 8 + 0])
+        avximagc = AVX.make_float8(imag_coords[i, j * 8 + 7],
+                                   imag_coords[i, j * 8 + 6],
+                                   imag_coords[i, j * 8 + 5],
+                                   imag_coords[i, j * 8 + 4],
+                                   imag_coords[i, j * 8 + 3],
+                                   imag_coords[i, j * 8 + 2],
+                                   imag_coords[i, j * 8 + 1],
+                                   imag_coords[i, j * 8 + 0])
+
+        avxiters = AVX.make_float8(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
+        avxones  = AVX.make_float8(1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0)
+        avxfours = AVX.make_float8(4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0) # to check for magnitude
+
+        # now do the loop
+        # init the series with c (we skip over the first step
+        # which is unnecessary in the original implemntation)
+        avxrealz = avxrealc
+        avximagz = avximagc
+
+        for iter in range(max_iterations):
+          # for a given complex number z = x + iy
+          # it holds z^2 = (x^2 - y^2) + i(2xy)
+          # we can use that to compute z^2 + c for 8 complex numbers in parallel!
+
+          # now one step
+          avxrealz_tmp = AVX.sub(AVX.mul(avxrealz, avxrealz), AVX.mul(avximagz, avximagz))
+          avximagz_tmp = AVX.mul(avxrealz, avximagz)
+          avximagz_tmp = AVX.add(avximagz_tmp, avximagz_tmp) # adding is faster and equal to * 2
+
```

Comment:
nice job creating temp variables to avoid overwriting data that you still need and using the wrong values